gene_expression_processing.py

```diff
@@ -0,0 +1,272 @@
+#!/usr/bin/env python3
+
+
+"""
+Hail Batch workflow to create gene expression files.
+This script will:
+
+- select genes to test based on expression (per cell type)
+- build chromosome & cell type specific phenotype covariate files
+- use gene info to create cis-window files
+- export pheno_cov files as tsv files
+- export cis_window files as tsv files
+
+More details in README
+output files in tob_wgs_genetics/saige_qtl/input
+"""
+
+import os
+import logging
+import math
+import sys
+
+import click
+import hail as hl
+import hailtop.batch as hb
+import pandas as pd
+import scanpy as sc
+
+from cpg_utils import to_path
+from cpg_utils.hail_batch import copy_common_env, dataset_path, output_path
+from cpg_workflows.batch import get_batch
+
+
+# use logging to print statements, display at info level
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s %(levelname)s %(module)s:%(lineno)d - %(message)s',
+    datefmt='%Y-%m-%d %H:%M:%S',
+    stream=sys.stderr,
+)
+
+CELLREGMAP_IMAGE = (
+    'australia-southeast1-docker.pkg.dev/cpg-common/images/cellregmap:0.0.1'
+)
+
+
+def filter_lowly_expressed_genes(expression_adata, min_pct=5) -> sc.AnnData:
+    """Remove genes with low expression across cells
+
+    Input: adata with all genes
+
+    Output: adata filtered
+    """
+    n_all_cells = len(expression_adata.obs.index)
+    min_cells = math.ceil((n_all_cells * min_pct) / 100)
+    expression_adata = sc.pp.filter_genes(expression_adata, min_cells=min_cells)
+    assert isinstance(expression_adata, sc.AnnData)
+
+    return expression_adata
+
+
+def get_chrom_celltype_expression(
+    gene_info_df,
+    expression_files_prefix: str,  # tob_wgs_genetics/saige_qtl/input/
+    chromosome: str,
+    cell_type: str,
+) -> sc.AnnData:
+    """Extracts relevant expression info
+
+    Input:
+    - chromosome & cell type of interest
+    - path to (single-cell) expression files, one tsv file per cell type,
+    rows are cells and columns are genes
+    - path to dataframe containing gene info, for each gene (=row),
+    specifies chrom, start, end and strand
+
+    Output: expression adata object for only relevant genes
+    """
+
+    # first line is where the file is now,
+    # but (second line) the files will eventually be in the below folder
+    # and split by cell type (at least this all naive B cells only)
+    expression_h5ad_path = to_path(
+        dataset_path(
+            f'scrna-seq/CellRegMap_input_files/expression_objects/sce{chromosome}.h5ad'
+        )
+    ).copy('here.h5ad')
```

Comment:
What does `.copy('here.h5ad')` do? It's listed twice on two separate lines

Reply:
One of these lines should be commented out. `copy(XXX)` does the cloudpathlib copy from remote to local (the scanpy library can't read remote paths). If/when this is hail batch'd this will be replaced with a `read_input` call